File: src/algorithms/strings/character-frequency/first-non-repeating-character/sources/first-non-repeating-character_test.py
"""Correctness tests for the first_non_repeating_character function."""

import importlib

module = importlib.import_module("first-non-repeating-character")
first_non_repeating_character = module.first_non_repeating_character


def test_leetcode():
    assert first_non_repeating_character("leetcode") == 0


def test_loveleetcode():
    assert first_non_repeating_character("loveleetcode") == 2


def test_aabb_all_repeat():
    assert first_non_repeating_character("aabb") == -1


def test_single_character():
    assert first_non_repeating_character("z") == 0


def test_aabbcc_all_repeat():
    assert first_non_repeating_character("aabbcc") == -1


def test_unique_in_middle():
    assert first_non_repeating_character("aabbc") == 4


def test_first_is_unique():
    assert first_non_repeating_character("xaabb") == 0


def test_last_is_unique():
    assert first_non_repeating_character("aabbz") == 4


def test_all_identical():
    assert first_non_repeating_character("aaaa") == -1


def test_two_unique_chars():
    assert first_non_repeating_character("ab") == 0


def test_dddccdbba():
    assert first_non_repeating_character("dddccdbba") == 8


if __name__ == "__main__":
    test_leetcode()
    test_loveleetcode()
    test_aabb_all_repeat()
    test_single_character()
    test_aabbcc_all_repeat()
    test_unique_in_middle()
    test_first_is_unique()
    test_last_is_unique()
    test_all_identical()
    test_two_unique_chars()
    test_dddccdbba()
    print("All tests passed!")

File: src/algorithms/strings/character-frequency/first-non-repeating-character/__tests__/first-non-repeating-character_test.go
package main

import "testing"

func TestFirstNonRepeatingCharacterLeetcode(t *testing.T) {
	if firstNonRepeatingCharacter("leetcode") != 0 {
		t.Error("expected 0 for 'leetcode'")
	}
}

func TestFirstNonRepeatingCharacterLoveleetcode(t *testing.T) {
	if firstNonRepeatingCharacter("loveleetcode") != 2 {
		t.Error("expected 2 for 'loveleetcode'")
	}
}

func TestFirstNonRepeatingCharacterAabbAllRepeat(t *testing.T) {
	if firstNonRepeatingCharacter("aabb") != -1 {
		t.Error("expected -1 for 'aabb'")
	}
}

func TestFirstNonRepeatingCharacterSingleChar(t *testing.T) {
	if firstNonRepeatingCharacter("z") != 0 {
		t.Error("expected 0 for 'z'")
	}
}

func TestFirstNonRepeatingCharacterAabbccAllRepeat(t *testing.T) {
	if firstNonRepeatingCharacter("aabbcc") != -1 {
		t.Error("expected -1 for 'aabbcc'")
	}
}

func TestFirstNonRepeatingCharacterUniqueInMiddle(t *testing.T) {
	if firstNonRepeatingCharacter("aabbc") != 4 {
		t.Error("expected 4 for 'aabbc'")
	}
}

func TestFirstNonRepeatingCharacterFirstIsUnique(t *testing.T) {
	if firstNonRepeatingCharacter("xaabb") != 0 {
		t.Error("expected 0 for 'xaabb'")
	}
}

func TestFirstNonRepeatingCharacterLastIsUnique(t *testing.T) {
	if firstNonRepeatingCharacter("aabbz") != 4 {
		t.Error("expected 4 for 'aabbz'")
	}
}

func TestFirstNonRepeatingCharacterAllIdentical(t *testing.T) {
	if firstNonRepeatingCharacter("aaaa") != -1 {
		t.Error("expected -1 for 'aaaa'")
	}
}

func TestFirstNonRepeatingCharacterTwoUniqueChars(t *testing.T) {
	if firstNonRepeatingCharacter("ab") != 0 {
		t.Error("expected 0 for 'ab'")
	}
}

func TestFirstNonRepeatingCharacterDddccdbba(t *testing.T) {
	if firstNonRepeatingCharacter("dddccdbba") != 8 {
		t.Error("expected 8 for 'dddccdbba'")
	}
}
